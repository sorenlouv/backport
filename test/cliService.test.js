const axios = require('axios');
const nock = require('nock');
const os = require('os');
const httpAdapter = require('axios/lib/adapters/http');

const cliService = require('../src/cli/cliService');
const rpc = require('../src/lib/rpc');
const commitMock = require('./mocks/commit.json');

axios.defaults.host = 'http://localhost';
axios.defaults.adapter = httpAdapter;

describe('doBackportVersion', () => {
  beforeEach(() => {
    os.homedir = jest.fn(() => '/homefolder');
    rpc.exec = jest.fn().mockReturnValue(Promise.resolve());
    rpc.mkdirp = jest.fn().mockReturnValue(Promise.resolve());

    this.addLabelMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/issues/1337/labels`, ['backport'])
      .query(true)
      .reply(200, {});
  });

  it('with pull request reference', () => {
    this.createPRMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/pulls`, {
        title: '[6.x] myCommitMessage | myOtherCommitMessage',
        body:
          'Backports the following commits to 6.x:\n - myCommitMessage (#myPullRequest)\n - myOtherCommitMessage (#myOtherPullRequest)',
        head: 'sqren:backport/6.x/pr-myPullRequest_pr-myOtherPullRequest',
        base: '6.x'
      })
      .query(true)
      .reply(200, {
        number: 1337,
        html_url: 'myHtmlUrl'
      });

    return cliService
      .doBackportVersion({
        owner: 'elastic',
        repoName: 'kibana',
        commits: [
          {
            sha: 'mySha',
            message: 'myCommitMessage',
            pullRequest: 'myPullRequest'
          },
          {
            sha: 'mySha2',
            message: 'myOtherCommitMessage',
            pullRequest: 'myOtherPullRequest'
          }
        ],
        branch: '6.x',
        username: 'sqren',
        labels: ['backport']
      })
      .then(res => {
        expect(res.config).toMatchSnapshot();
        expect(rpc.exec.mock.calls).toMatchSnapshot();
        expect(this.createPRMock.isDone()).toBe(true);
        expect(this.addLabelMock.isDone()).toBe(true);
      });
  });

  it('without pull request reference', () => {
    this.createPRMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/pulls`, {
        title: '[6.x] myCommitMessage',
        body:
          'Backports the following commits to 6.x:\n - myCommitMessage (mySha)',
        head: 'sqren:backport/6.x/commit-mySha',
        base: '6.x'
      })
      .query(true)
      .reply(200, {
        number: 1337,
        html_url: 'myHtmlUrl'
      });

    return cliService
      .doBackportVersion({
        owner: 'elastic',
        repoName: 'kibana',
        commits: [
          {
            sha: 'mySha',
            message: 'myCommitMessage'
          }
        ],
        branch: '6.x',
        username: 'sqren'
      })
      .then(res => {
        expect(res.config).toMatchSnapshot();
        expect(rpc.exec.mock.calls).toMatchSnapshot();
        expect(this.createPRMock.isDone()).toBe(true);
        expect(this.addLabelMock.isDone()).toBe(false);
      });
  });
});

describe('getCommitBySha', () => {
  beforeEach(() => {
    nock('https://api.github.com')
      .get(`/repos/elastic/kibana/commits/mySha`)
      .query(true)
      .reply(200, commitMock);

    return cliService
      .getCommitBySha({
        owner: 'elastic',
        repoName: 'kibana',
        sha: 'mySha'
      })
      .then(commits => (this.commits = commits));
  });

  it('should return a single commit in an array', () => {
    expect(this.commits).toEqual([
      {
        message: '[Chrome] Bootstrap Angular into document.body (#15158)',
        sha: 'f3430595978a6123c65f7501e61386de62b80b6e'
      }
    ]);
  });
});

describe('getReferenceLong', () => {
  it('should return a sha', () => {
    expect(cliService.getReferenceLong({ sha: 'mySha1234567' })).toEqual(
      'mySha12'
    );
  });

  it('should return a pr', () => {
    expect(
      cliService.getReferenceLong({ pullRequest: '1337', sha: 'mySha1234567' })
    ).toEqual('#1337');
  });
});

describe('withPullRequest', () => {
  function mockGithubIssuesResponse(res) {
    return nock('https://api.github.com')
      .get(`/search/issues`)
      .query(true)
      .reply(200, res);
  }

  it('should decorate commit with pull request if available', () => {
    mockGithubIssuesResponse({
      items: [{ number: 'myPullRequest' }]
    });

    return cliService
      .withPullRequest('elastic', 'kibana', [{ message: 'myCommitMessage' }])
      .then(res =>
        expect(res).toEqual([
          { message: 'myCommitMessage', pullRequest: 'myPullRequest' }
        ])
      );
  });

  it('should not decorate commit, when pull request is not available', () => {
    mockGithubIssuesResponse({ items: [] });

    return cliService
      .withPullRequest('elastic', 'kibana', [
        { sha: 'myCommitSha', message: 'myCommitMessage' }
      ])
      .then(res =>
        expect(res).toEqual([
          {
            sha: 'myCommitSha',
            message: 'myCommitMessage',
            pullRequest: undefined
          }
        ])
      );
  });
});
