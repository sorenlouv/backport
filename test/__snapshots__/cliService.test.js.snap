// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doBackportVersion with pull request reference 1`] = `
Object {
  "adapter": [Function],
  "data": "{\\"title\\":\\"[6.x] myCommitMessage | myOtherCommitMessage\\",\\"body\\":\\"Backports the following commits to 6.x:\\\\n - myCommitMessage (#myPullRequest)\\\\n - myOtherCommitMessage (#myOtherPullRequest)\\",\\"head\\":\\"sqren:backport/6.x/pr-myPullRequest_pr-myOtherPullRequest\\",\\"base\\":\\"6.x\\"}",
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Content-Length": 269,
    "Content-Type": "application/json;charset=utf-8",
    "User-Agent": "axios/0.17.1",
  },
  "host": "http://localhost",
  "maxContentLength": -1,
  "method": "post",
  "timeout": 0,
  "transformRequest": Object {
    "0": [Function],
  },
  "transformResponse": Object {
    "0": [Function],
  },
  "url": "https://api.github.com/repos/elastic/kibana/pulls?access_token=undefined",
  "validateStatus": [Function],
  "xsrfCookieName": "XSRF-TOKEN",
  "xsrfHeaderName": "X-XSRF-TOKEN",
}
`;

exports[`doBackportVersion with pull request reference 2`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/pr-myPullRequest_pr-myOtherPullRequest origin/6.x --force && git checkout backport/6.x/pr-myPullRequest_pr-myOtherPullRequest ",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha2",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/pr-myPullRequest_pr-myOtherPullRequest --force",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
]
`;

exports[`doBackportVersion without pull request reference 1`] = `
Object {
  "adapter": [Function],
  "data": "{\\"title\\":\\"[6.x] myCommitMessage\\",\\"body\\":\\"Backports the following commits to 6.x:\\\\n - myCommitMessage (mySha)\\",\\"head\\":\\"sqren:backport/6.x/commit-mySha\\",\\"base\\":\\"6.x\\"}",
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Content-Length": 164,
    "Content-Type": "application/json;charset=utf-8",
    "User-Agent": "axios/0.17.1",
  },
  "host": "http://localhost",
  "maxContentLength": -1,
  "method": "post",
  "timeout": 0,
  "transformRequest": Object {
    "0": [Function],
  },
  "transformResponse": Object {
    "0": [Function],
  },
  "url": "https://api.github.com/repos/elastic/kibana/pulls?access_token=undefined",
  "validateStatus": [Function],
  "xsrfCookieName": "XSRF-TOKEN",
  "xsrfHeaderName": "X-XSRF-TOKEN",
}
`;

exports[`doBackportVersion without pull request reference 2`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/commit-mySha origin/6.x --force && git checkout backport/6.x/commit-mySha ",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/commit-mySha --force",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
]
`;
